test/framework/controlplane_helpers.go

```diff
@@ -164,20 +163,34 @@ type WaitForControlPlaneToBeReadyInput struct {
 func WaitForControlPlaneToBeReady(ctx context.Context, input WaitForControlPlaneToBeReadyInput, intervals ...interface{}) {
 	By("Waiting for the control plane to be ready")
 	controlplane := &controlplanev1.KubeadmControlPlane{}
-	Eventually(func() (controlplanev1.KubeadmControlPlane, error) {
+	Eventually(func() (bool, error) {
 		key := client.ObjectKey{
 			Namespace: input.ControlPlane.GetNamespace(),
 			Name:      input.ControlPlane.GetName(),
 		}
 		if err := input.Getter.Get(ctx, key, controlplane); err != nil {
-			return *controlplane, errors.Wrapf(err, "failed to get KCP")
+			return false, errors.Wrapf(err, "failed to get KCP")
+		}
+
+		desiredReplicas := controlplane.Spec.Replicas
+		statusReplicas := controlplane.Status.Replicas
+		updatedReplicas := controlplane.Status.UpdatedReplicas
+		readyReplicas := controlplane.Status.ReadyReplicas
+		unavailableReplicas := controlplane.Status.UnavailableReplicas
+
+		// Control plane is still rolling out (and thus not ready) if:
+		// * .spec.replicas, .status.replicas, .status.updatedReplicas,
+		//   .status.readyReplicas are not equal and
+		// * unavailableReplicas > 0
+		if statusReplicas != *desiredReplicas ||
+			updatedReplicas != *desiredReplicas ||
+			readyReplicas != *desiredReplicas ||
+			unavailableReplicas > 0 {
+			return false, nil
 		}
```

Comment:
We do this because:
* KCP is only stable if all of those conditions are fulfilled
* By waiting for the same conditions in the test code and in the topology controller we can ensure that setting etcdImageTag and the version in Cluster.spec.topology (in the cluster upgrade test) doesn't trigger a double rollout.
 This is because when both the version and the variable are set the topology controller will roll them out at the same time and not hold back the version because KCP is not stable.